src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -1,79 +1,260 @@
+  private def handleStreamException(ctx: RequestContext, t: Throwable): Unit = {
+    val message = t match {
+      case f: FireCloudExceptionWithErrorReport => s"FireCloudException: Error generating entity download: ${f.errorReport.message}"
+      case _ => s"FireCloudException: Error generating entity download: ${t.getMessage}"
+    }
+    ctx.responder ! MessageChunk(message)
+    ctx.responder ! ChunkedMessageEnd
+  }
+
+  /**
+    * General Approach
+    * 1. Define a `Source` of entity queries
+    * 2. Run the source events through a `Flow`.
+    * 3. Flow sends events (batch of entities) to a streaming output actor
+    * 4. Return a Done to the calling route when complete.
+    */
+  private def streamSingularType(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, headers: IndexedSeq[String], attributeNames: Option[IndexedSeq[String]]): Future[Done] = {
+    // The output to the user
+    val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`text/plain`, entityType + ".txt")))
+
+    // The Source
+    val entityQuerySource = Source(entityQueries.toStream)
+
+    // The Flow. Using mapAsync(1) ensures that we run 1 batch at a time through this side-affecting process.
+    val flow = Flow[EntityQuery].mapAsync(1) { query =>
+      getEntitiesFromQuery(workspaceNamespace, workspaceName, entityType, query) map { entities =>
+        sendRowsAsChunks(streamingActorRef, query, entityQueries.size, entityType, headers, entities)
       }
     }
+
+    // Ignore the result - we don't need to remember anything about this operation.
+    val sink = Sink.ignore
+
+    // finally, run it:
+    entityQuerySource.via(flow).runWith(sink)
   }
 
-  private def getZipBytes(entityType: String, membershipData: String, entityData: String): Array[Byte] = {
-    val bos = new ByteArrayOutputStream()
-    val zos = new ZipOutputStream(bos)
-    zos.putNextEntry(new ZipEntry(entityType + "_membership.tsv"))
-    zos.write(membershipData.getBytes)
-    zos.closeEntry()
-    zos.putNextEntry(new ZipEntry(entityType + "_entity.tsv"))
-    zos.write(entityData.getBytes)
-    zos.closeEntry()
-    zos.finish()
-    bos.toByteArray
+  private def streamCollectionType(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, entityType: String, entityQueries: Seq[EntityQuery], metadata: EntityTypeMetadata, attributeNames: Option[IndexedSeq[String]]): Future[Done] = {
+
+    // Future of all entities
+    val entityBatches = Future.traverse(entityQueries) { query =>
+      getEntitiesFromQuery(workspaceNamespace, workspaceName, entityType, query)
+    } map (_.flatten)
+
+    // Two File sinks, one for each kind of entity set file needed.
+    // The temp files will end up zipped and streamed when complete.
+    val tempEntityFile: File = File.newTemporaryFile()
```

Comment:
Worth giving these tmp files a prefix? Like:
```
val tempEntityFile: File = File.newTemporaryFile(prefix = "entity_")
```
It might help with debugging down the road.